### ISSUE TITLE
tf.keras.ops.numpy.ldexp returns incorrect gradient at zero

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#127244]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13.5

### Current behavior?

`tf.keras.ops.numpy.ldexp` returns an incorrect first derivative for a finite `float64` input.

In the reproducer below, the integer exponents are fixed and broadcast over the input. Therefore, `ldexp(x, e)` is just scaling by fixed powers of two, and the weighted result is affine in `t`. Zero is not a singularity of scaling by a fixed power of two.

The forward value is computed correctly as `101.5`, but reverse-mode autodiff returns `209.0` for the derivative with respect to `t`. The expected derivative is `253.125`.

Expected behavior?

The first derivative of the weighted `tf.keras.ops.numpy.ldexp` expression at `t = 0.0` should be close to `253.125`, not `209.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    x = tf.reshape(tf.stack([t, 2 * t + 1]), (2, 1))
    e = tf.constant([-3, 0, 4], dtype=tf.int32)
    y = tf.keras.ops.numpy.ldexp(x, e)
    return sum((1 + 3 * i + j) * y[i, j] for i in range(2) for j in range(3))

x = tf.constant(0.0, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 253.125

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.keras.ops.numpy.ldexp returns incorrect gradient at zero")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: 101.5
actual: 209.0
expected: 253.125
BUG REPRODUCED: tf.keras.ops.numpy.ldexp returns incorrect gradient at zero
```

**Comment by jdymitarai** [comment]
I have identified the root cause and submitted a fix in **keras-team/keras#23615**.

### Root Cause Analysis
In `keras/src/backend/tensorflow/numpy.py:2017`, `ldexp` was defined as:
```python
tf.where(tf.math.is_inf(x1) | (x1 == 0), x1, result)
```
When `x1 == 0`, `tf.where` routes directly to `x1`. Under reverse-mode automatic differentiation, evaluating the Jacobian/derivative with respect to `x1` yields:
$$\frac{\partial}{\partial x_1}(x_1) = 1.0$$
instead of the correct mathematical derivative:
$$\frac{\partial}{\partial x_1}(x_1 \cdot 2^{x_2}) = 2^{x_2}$$

For instance, when $x_1 = 0$ and $x_2 = 5$, the gradient should be $2^5 = 32.0$, but was returning $1.0$.

### Resolution
$0 \cdot 2^{x_2} = 0$ is naturally well-defined in floating point arithmetic and does not produce `NaN`. Only $\pm\infty$ requires special masking to avoid indeterminate $\infty \times 0$. Removing `| (x1 == 0)` preserves exact mathematical gradients across all inputs while keeping $\pm\infty$ safe.

Fixed in https://github.com/keras-team/keras/pull/23615 with comprehensive unit tests for zero-input reverse-mode autodiff gradients.
